### DISCUSSION TITLE
Odd behavior when using or_ function with collection object

### DISCUSSION THREAD

**Discussion by jordantshaw** [sqlalchemy/sqlalchemy#7840] [Usage Questions]
SqlAlchemy: 1.4.32 
Python: 3.9.10
Driver: Pyodbc/MSSQL

I am seeing odd behavior when using the `or_` function in a query filter.  When using the `or_` function to filter on collection relationship objects combined with standard columns, SQLAlchemy seems to just freeze and never return a query result. This does not happen if I used `and_`.  It also does not happen if I only filter on standard columns or if I only filter on collection type columns.  But as soon as I mix a collection column with a standard column in the filter, it never returns a result from the query being executed.

Sample:
```
    query = (
        select(models.Person)
        .filter(or_(
            models.Person.phone_numbers.any(models.EntityPhone.phone_number.contains(search_value)),
            models.Person.emails.any(models.Emails.email_address.contains(search_value)),
            models.Person.short_name.contains(search_value)
        ))
    )

  results = session.execute(query).scalars().all()
```

In the above example when executing the query no results are ever returned, no error is raised, the code just freezes at that point until I eventually stop the script. If I change the  query and remove the last filter it runs with no issues (see example below).

```
    query = (
        select(models.Person)
        .filter(or_(
            models.Person.phone_numbers.any(models.EntityPhone.phone_number.contains(search_value)),
            models.Person.emails.any(models.Emails.email_address.contains(search_value)),
        ))
    )

  results = session.execute(query).scalars().all()
```

When I set echo=True in the engine, the query is generated correctly, and I can even copy that query and run it in an external SQL client, and it runs just fine with no issues.  Only when executing from SQLAlachemy does it seem to hang indefinitely.

Any ideas on what could be causing this?

**Comment by CaselIT** [comment]
Hi,

This seems like the db is taking a lot of time or the number of results it very large (so sqlalchemy is taking time to create the objects).
How long did you wait for the query to complete?
Is the table you are querying very large?

**Reply by jordantshaw** [reply]
@CaselIT  that was my initial thought as well.  I have let it run for upwards of 10 mins before killing it.  However, like I mentioned, I run the generated query manually, it returns the results in just a few milliseconds.  It only returns 1 row.  The table only contains about 100k records.

**Reply by zzzeek** [reply]
youd want to look at all the SQL being logged, are there lazy loads going on, or a "subqueryload", etc.  is it defintiely the "good" query that it's hanging on ?  if you press ctrl-C, where does the stack trace show where it was ?

**Reply by jordantshaw** [reply]
@zzzeek I did some more digging.  Later on I was adding a `limit()` function to this query.  That is what is causing this query to hang.  So if no limit is used, it returns a few milliseconds, if I add limit, it just hangs until I kill the query.  Any thoughts?

`query.limit(pagination.limit + 1)`

**Reply by zzzeek** [reply]
if you are using orm joinedload(), the limit() can turn the whole query into a SELECT from a subquery in order to apply the limit. You'd see this in the echo=True log.

**Reply by jordantshaw** [reply]
@zzzeek So I looked into the query plans from MSSQL.  Looks like, for some weird reason, when the TOP clause is added to the query, SQLServer uses a very inefficient execution plan.  Definitely seems like a SQL Server issue.  I was able to add some suggested indexes and it resolved the issue.

**Reply by zzzeek** [reply]
that TOP would have been in the SQL logs that you saw when you did echo=True.   did you originally run the exact SQL that was logged manually to confirm it didn't have a speed issue?

**Reply by jordantshaw** [reply]
I neglected to include the TOP when I ran the query manually, as I didnt think anything of it. Really odd that TOP would effect SQL execution plans. When I added the top to my manual query I replicated the same issue.

**Reply by CaselIT** [reply]
If your mssql version support it you may also use fetch first instead of top (but I don't remember if the compiler is made to prefer top vs fetch in mssql) https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=fetch#sqlalchemy.sql.expression.Select.fetch
